modules/imgproc/src/color_rgb.cpp

```diff
@@ -133,107 +133,101 @@ struct RGB5x52RGB
 
     RGB5x52RGB(int _dstcn, int _blueIdx, int _greenBits)
         : dstcn(_dstcn), blueIdx(_blueIdx), greenBits(_greenBits)
-    {
-        #if CV_NEON
-        v_n3 = vdupq_n_u16(~3);
-        v_n7 = vdupq_n_u16(~7);
-        v_255 = vdupq_n_u8(255);
-        v_0 = vdupq_n_u8(0);
-        v_mask = vdupq_n_u16(0x8000);
-        #endif
-    }
+    { }
 
     void operator()(const uchar* src, uchar* dst, int n) const
     {
-        int dcn = dstcn, bidx = blueIdx, i = 0;
-        if( greenBits == 6 )
+        int dcn = dstcn, bidx = blueIdx, gb = greenBits;
+        int i = 0;
+
+#if CV_SIMD
+        const int vsize = v_uint8::nlanes;
+        v_uint8 vz = vx_setzero_u8(), vn0 = vx_setall_u8(255);
+        for(; i < n-vsize+1;
+            i += vsize, src += vsize*sizeof(ushort), dst += vsize*dcn)
         {
-            #if CV_NEON
-            for ( ; i <= n - 16; i += 16, dst += dcn * 16)
+            v_uint16 t0 = v_reinterpret_as_u16(vx_load(src));
+            v_uint16 t1 = v_reinterpret_as_u16(vx_load(src +
+                                                       sizeof(ushort)*v_uint16::nlanes));
+
+            //TODO: shorten registers use when v_interleave is available
+            v_uint8 r, g, b, a;
+            v_uint16 b0 = (t0 << 11) >> 8;
+            v_uint16 b1 = (t1 << 11) >> 8;
+            b = v_pack(b0, b1);
+
+            v_uint16 g0, g1, r0, r1, a0, a1;
+
+            if( gb == 6 )
+            {
+                g0 = ((t0 >> 5) << 10) >> 8;
```

Comment:
Wouldn't one bit-mask be faster than two shifts?

Reply:
I did exactly that in the beginng then decided to use shifts instead since masking requres one extra register for mask.
Can't state which variant works faster. Their performance differs less than between any two runs of identical code.